doc/manual/ml/kmeans.3
.TH "kmeans" "" "" "" ""
.SH NAME
.PP
kmeans \- A clustering algorithm commonly used in EDA (exploratory data
analysis).
.SH SYNOPSIS
.PP
\f[C]#include\ <frovedis/ml/clustering/kmeans.hpp>\f[]
.PP
\f[C]rowmajor_matrix_local<T>\f[]
.PD 0
.P
.PD
frovedis::kmeans (\f[C]crs_matrix<T,I,O>\f[]& samples,
.PD 0
.P
.PD
\  \  \  \ int k,
.PD 0
.P
.PD
\  \  \  \ int iter,
.PD 0
.P
.PD
\  \  \  \ T eps,
.PD 0
.P
.PD
\  \  \  \ long seed = 0)
.PP
\f[C]std::vector<int>\f[]
.PD 0
.P
.PD
frovedis::kmeans_assign_cluster (\f[C]crs_matrix_local<T,I,O>\f[]& mat,
.PD 0
.P
.PD
\  \  \  \  \  \  \  \ \f[C]rowmajor_matrix_local<T>\f[]& centroid)
.SH DESCRIPTION
.PP
Clustering is an unsupervised learning problem whereby we aim to group
subsets of entities with one another based on some notion of similarity.
K\-means is one of the most commonly used clustering algorithms that
clusters the data points into a predefined number of clusters (K).
.SS Detailed Description
.SS frovedis::kmeans()
.PP
\f[B]Parameters\f[]
.PD 0
.P
.PD
\f[I]samples\f[]: A \f[C]crs_matrix<T,I,O>\f[] containing the sparse
data points
.PD 0
.P
.PD
\f[I]k\f[]: An integer parameter containing the number of clusters
.PD 0
.P
.PD
\f[I]iter\f[]: An integer parameter containing the maximum number of
iteration count
.PD 0
.P
.PD
\f[I]eps\f[]: A parameter of T type containing the epsilon value
.PD 0
.P
.PD
\f[I]seed\f[]: A parameter of long type containing the seed value to
generate the random rows from the given data samples (Default: 0)
.PP
\f[B]Purpose\f[]
.PD 0
.P
.PD
It clusters the given data points into a predefined number (k) of
clusters.
.PD 0
.P
.PD
After the successful clustering, it returns the k centroids of the
cluster.
.PP
\f[B]Return Value\f[]
.PD 0
.P
.PD
After the successful ustering it returns the centroids of the type
\f[C]rowmajor_matrix_local<T>\f[], where each column shows each centroid
vector.
.SS frovedis::kmeans_assign_cluster()
.PP
\f[B]Parameters\f[]
.PD 0
.P
.PD
\f[I]mat\f[]: A \f[C]crs_matrix_local<T,I,O>\f[] containing the new
sparse data points to be assigned to the cluster
.PD 0
.P
.PD
\f[I]centroid\f[]: A \f[C]rowmajor_matrix_local<T>\f[] contaning the
centroids
.PP
\f[B]Purpose\f[]
.PD 0
.P
.PD
After getting the centroids from kmeans(), they can be used to assign
data to the closest centroid using kmeans_assign_cluster().
.PP
\f[B]Return Value\f[]
.PD 0
.P
.PD
It returns a \f[C]std::vector<int>\f[] containing the assigned values.


samples/kmeans/kmeans.cc
#include <frovedis.hpp>
#include <frovedis/ml/clustering/kmeans.hpp>

#include <boost/program_options.hpp>

using namespace boost;
using namespace frovedis;
using namespace std;

void do_kmeans(const string& input, const string& output, int k,
               int num_iteration, double eps, bool binary) {
  if(binary) {
    time_spent t(DEBUG);
    auto mat = make_crs_matrix_loadbinary<double>(input);
    t.show("load matrix: ");
    auto r = kmeans(std::move(mat), k, num_iteration, eps);
    t.show("kmeans time: ");
    r.transpose().savebinary(output);
    t.show("save centroid time: ");
  } else {
    time_spent t(DEBUG);
    auto mat = make_crs_matrix_load<double>(input);
    t.show("load matrix: ");
    auto r = kmeans(std::move(mat), k, num_iteration, eps);
    t.show("kmeans time: ");
    r.transpose().save(output);
    t.show("save model time: ");
  }
}

void do_assign(const string& input, const string& input_centroid,
               const string& output, bool binary) {
  if(binary) {
    time_spent t(DEBUG);
    auto mat = make_crs_matrix_local_loadbinary<double>(input);
    t.show("load matrix: ");
    auto c = make_rowmajor_matrix_local_loadbinary<double>(input_centroid);
    t.show("load centroid: ");
    auto ct = c.transpose();
    auto r = kmeans_assign_cluster(mat, ct);
    make_dvector_scatter(r).savebinary(output);    
  } else {
    time_spent t(DEBUG);
    auto mat = make_crs_matrix_local_load<double>(input);
    t.show("load matrix: ");
    auto c = make_rowmajor_matrix_local_load<double>(input_centroid);
    t.show("load centroid: ");
    auto ct = c.transpose();
    auto r = kmeans_assign_cluster(mat, ct);
    make_dvector_scatter(r).saveline(output);    
  }
}

int main(int argc, char* argv[]) {
  use_frovedis use(argc, argv);

  using namespace boost::program_options;

  options_description opt("option");
  opt.add_options()
    ("help,h", "print help")
    ("assign,a", "assign data to cluster mode")
    ("input,i", value<string>(), "input matrix")
    ("centroid,c", value<string>(), "input centroid for assignment")
    ("output,o", value<string>(), "output centroids or cluster")
    ("k,k", value<int>(), "number of clusters")
    ("num-iteration,n", value<int>(), "number of max iteration")
    ("eps,e", value<double>(), "epsilon to stop the iteration")
    ("sparse,s", "use sparse matrix [default]")
    ("dense,d", "use dense matrix [not supported yet]")
    ("verbose", "set loglevel to DEBUG")
    ("verbose2", "set loglevel to TRACE")
    ("binary,b", "use binary input/output");

  variables_map argmap;
  store(command_line_parser(argc,argv).options(opt).allow_unregistered().
        run(), argmap);
  notify(argmap);

  string input, output, input_centroid;
  int num_iteration = 100;
  int k;
  double eps = 0.01;
  bool assign = false;
  bool dense = false;
  bool binary = false;
  
  if(argmap.count("help")){
    cerr << opt << endl;
    exit(1);
  }

  if(argmap.count("input")){
    input = argmap["input"].as<string>();
  } else {
    cerr << "input is not specified" << endl;
    cerr << opt << endl;
    exit(1);
  }

  if(argmap.count("output")){
    output = argmap["output"].as<string>();
  } else {
    cerr << "output is not specified" << endl;
    cerr << opt << endl;
    exit(1);
  }

  if(argmap.count("assign")){
    assign = true;
  }

  if(argmap.count("centroid")){
    input_centroid = argmap["centroid"].as<string>();
  } else {
    if(assign == true) {
      cerr << "output is not specified" << endl;
      cerr << opt << endl;
      exit(1);
    }
  }

  if(argmap.count("k")){
    k = argmap["k"].as<int>();
  } else {
    if(assign == false) {
      cerr << "number of cluster is not specified" << endl;
      cerr << opt << endl;
      exit(1);
    }
  }

  if(argmap.count("num-iteration")){
    num_iteration = argmap["num-iteration"].as<int>();
  }

  if(argmap.count("epsilon")){
    eps = argmap["epsilon"].as<double>();
  }

  if(argmap.count("sparse")){
    dense = false;
  }

  if(argmap.count("dense")){
    dense = true;
  }

  if(argmap.count("binary")){
    binary = true;
  }

  if(argmap.count("verbose")){
    set_loglevel(DEBUG);
  }

  if(argmap.count("verbose2")){
    set_loglevel(TRACE);
  }

  if(dense) {
    cerr << "currently dense matrix is not supported" << endl;
    exit(1);
  } else {
    if(assign) do_assign(input, input_centroid, output, binary);
    else do_kmeans(input, output, k, num_iteration, eps, binary);
  }
}
